### DISCUSSION TITLE
Using Route classes instead of hash_branches

### DISCUSSION THREAD

**Discussion by RomanTurner** [jeremyevans/roda#305] [General]
Hey there, 

I was working on setting up my web framework using Roda, Zeitwerk, and Dry-System. 
I was running into a lot of issues getting around code reloading and have been problem-solving this for a couple of weeks at this point when in conversation with another member of the community, we'd discussed mounting what was hash_branches as their own Roda applications. This worked! Zeitwerk is happy, and I haven't encountered any problems when I manually climb them. 

That being said, there is probably a performance hit by mounting the apps like that compared to hash_branches. Still, I've been pulling my hair out on this code-loading issue for long enough that I am okay with it until a spark of inspiration or some help can come my way. 

**The discussion question: Are there drawbacks to mounting and running individual Roda apps for each branched route?** 

Some more details below highlight the directory structure and how I am metaprogramming the apps and mounting them. 

The directory structure is not the traditional "MVC" or bucket system. It is based on allocating resources and the locality of behavior. It looks like such:

```
project
│   README.md
│   app.rb
|   config.ru
│
└───allocs
│   │
│   └─── users
│       │   data_model.rb
│       │   routes.rb
│       │   name_decorator.rb
│   │
│   └─── posts
│       │   data_model.rb
│       │   routes.rb
│       │   publisher.rb
|
└───system
```

With some metaprogramming and directory conventions, there could be a route that loads all the subpath applications themselves. There might be some edge cases on constating; if you have any thoughts, feel free to let me know. 

```ruby
    inflector = Dry::Inflector.new
    Dir['allocs/*'].each do |file|
      route_name = File.basename(file)
      r.on(route_name) do
        klass = inflector.classify(route_name)
        r.run inflector.constantize("#{klass}::Routes")
      end
    end
```

**Comment by jeremyevans** [comment]
Using separate Roda applications for routing branches is fine.  It's slightly slower than the `hash_branches` approach, but probably not enough to matter in a production application.

For this approach, you could consider using the `multi_run` plugin that ships with Roda.  In `multi_run`'s current state, it would probably cause eager loading of all the routing branches, but that could probably be avoided with some changes to the plugin so that the constants aren't referenced until the routes are accessed (one idea: `run("branch"){Branch::Routes}`).

Note that if you want code reloading for an application that uses hash_branches, you can use rack-unreloader for the autoloading/reloading part.  That's the approach I use in my production applications, and how roda-sequel-stack is setup (https://github.com/jeremyevans/roda-sequel-stack).  But I can understand wanting to use Zeitwerk instead, and the `hash_branches` approach doesn't work well with Zeitwerk due to Zeitwerk's limitations.

**Reply by RomanTurner** [reply]
Thanks for such a quick reply. I tried to set up rack-unreloader and was stymied, unfortunately. I've combed over the roda-sequel-stack project and have borrowed so much from it to set up my own applications. I am learning a lot, and I think there will be a time when it will 'click,' but right now I have spent a lot of time without making much visible progress, and it is getting to me 😄

I'll look at the `multi-run` plugin and try to avoid the eager loading in development mode. On that note, I enjoy the autoloading hash_branches plugin, so having similar behavior would boost larger applications.

**Comment by mmarvb8h** [comment]
I know this may be a little late. But i was working on this last night. I followed ur example with hash-branches and sure enough Zeitwerk complained. What i have done so far is to follow the project structure, but have Zeitwerk exclude the routes directory. I use the autoload_hash_branches plugin to load the routes directory. Seems to work fine.

I have not tried Zeitwerk reloading yet since i have such a small prototype app right now and its just as easy to restart. I'll be trying the autoload in a week or so.
